lxd/db/openfga/openfga.go

```diff
@@ -205,6 +219,79 @@ func (o *openfgaStore) ReadUserTuple(ctx context.Context, store string, tk *open
 	return nil, nil
 }
 
+func (o *openfgaStore) ensureCacheLoaded(ctx context.Context, cache *RequestCache) error {
+	// Check if already loaded.
+	if cache.initialised.Load() {
+		return nil
+	}
+
+	// If not loaded, lock the cache and set loaded to true. This should mean that concurrent callers may think the cache
+	// is loaded when it isn't yet - but it will be loaded when they are able to acquire a lock.
+	cache.permissionsByEntityTypeMu.Lock()
+	cache.permissionsByGroupMu.Lock()
+	defer cache.permissionsByEntityTypeMu.Unlock()
+	defer cache.permissionsByGroupMu.Unlock()
+	cache.initialised.Store(true)
+
+	// Get a map of group to slice of permissions.
+	var groupPermissions map[string][]cluster.Permission
+	err := o.clusterDB.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+		var err error
+		groupPermissions, err = cluster.GetGroupPermissions(ctx, tx.Tx())
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	// Rearrange our map to optimise for ReadUsersetTuples and ReadStartingWithUser.
+	// - ReadUsersetTuples asks the question: Which groups have this permission?
+	//   So it essentially maps permissions to a list of groups. This is the permissionCacheByEntityType.
+	// - ReadStartingFromUser asks the question: Which entities does this group have this entitlement on?
+	//   So it essentially maps groups to permissions. This is the permisionCacheByGroup.
+	permissionCacheByEntityType := make(map[entity.Type]map[auth.Entitlement]map[int][]string)
```

Comment:
Could you please move these descriptions to their definition in the OpenFGA cache struct?

Reply:
fixed